### PR TITLE
Improve convergence of rk4 particle advection in time-dependent flow

### PR DESCRIPTION
**Pull request by gassmoeller**
This PR applies the same time interpolation we use for the RK2 integrator to the RK4 scheme. This way we achieve second order convergence in time variable flow fields. The renaming of variables is in line with the naming scheme of the RK2 class.

### CODE REVIEW

File: source/particle/integrator/rk_4.cc [hunk 1]
```diff
@@ -52,32 +52,32 @@ namespace aspect
...
               {
                 loc0[particle_id] = it->second.get_location();
-                k1[particle_id] = dt * (*vel);
+                k1[particle_id] = dt * (*old_velocity);
                 it->second.set_location(it->second.get_location() + 0.5*k1[particle_id]);
               }
             else if (integrator_substep == 1)
               {
-                k2[particle_id] = dt * (*vel);
+                k2[particle_id] = dt * (*old_velocity + *velocity) / 2.0;
```
Comment (bangerth) [inline source/particle/integrator/rk_4.cc]: Yes, that is the change we talked about in last week's group meeting.

**Comment by bangerth** [comment]
This is strict progress, so please merge if the tester is happy.

I think there is still some progress possible by also taking into account `old_old_velocity` for time steps >2. You would have to interpolate to time `t_{n+1/2}` based on `u^{n-1}, u^n, u^{n+1}`. This may improve the convergence order by one more.

**Comment by gassmoeller** [comment]
I will wait for the tester before merging. Concerning the use of more timesteps: I will check that at some point, but I am hesitant to move forward with that before thinking about our time stepping scheme again for active particles. Currently we do particle advection after the Stokes solution, which is not strictly correct for active particles. To be correct we might need to do something like the following:
Move the particle advection step before or after the solution of the advection-difffusion equations, use the old and old old solution to generate an extrapolated velocity and use that instead of the current method. We might also be able to use the current linearization point to advect particles better in nonlinear flow. But we indeed might be able to gain another order by your suggestion.

**Comment by bangerth** [comment]
Well, it's an operator splitting scheme: you have to either advance the velocity first and the particles second, or the other way around. There is nothing much you can do about it. For the coupling between velocity and advected quantities, one can think of a coupled scheme via a Newton method, but I really don't think we want to go to this kind of length with coupling particles :-)

**Comment by gassmoeller** [comment]
I was not thinking of something quite so complicated. But I would prefer to make the particle behavior as similar to the fields as possible to prevent confusions and that would mean solving both *before* the velocity instead of one before and the other after the velocity. This would also allow something like iterated impes for particle movement.

**Comment by gassmoeller** [comment]
I have opened #829 for discussion and will merge this PR for now.

**Comment by egpuckett** [comment]
I've had the task of figuring out how to make RK4 truly fourth order in time.  Should I open an issue or just monkey around with my ideas until I either get something or not?

**Comment by bangerth** [comment]
Opening issues for discussion that can later be closed by a patch is always a good idea.

**Comment by egpuckett** [comment]
On 04/23/2016 03:35 PM, Wolfgang
      Bangerth wrote:
    
    
      Opening issues for discussion that can later be closed by a
        patch is always a good idea. 
      —
        You are receiving this because you commented.
        Reply to this email directly or view
          it on GitHub
      
        
          
          
        
        
      
    
    I opened #830.
    
      
        On 04/23/2016 03:36 PM, Wolfgang
          Bangerth wrote:
        
        
          We hear and see everything, Gerry!
        
      
    
    I'm sure you've seen it by now!  I don't know how to add a label
    such as 
    
      
      e enhancement

**Pull request merged**